src/Journal.cpp

```diff
@@ -16,19 +16,486 @@
+}
+
+// When playing, find a window by path, corresponding to the window that had the
+// same path in a previous run
+wxWindow *FindByPath( const Path &path )
+{
+   auto components = wxSplit( path.GET(), PathSeparator, EscapeCharacter );
+   if ( !components.empty() ) {
+      auto iter = components.begin(), end = components.end();
+      auto pWindow = FindByNameAmongPeers( *iter++, wxTopLevelWindows );
+      while ( pWindow && iter != end )
+         pWindow = FindByNameAmongPeers( *iter++, pWindow->GetChildren() );
+      return pWindow;
+   }
+   return nullptr;
+}
+
+}
+
+namespace Events {
+
+/*!
+utilities to record events to journal and recreate them on playback
+*/
+
+// Events need to be recorded in the journal, but the numbers associated with
+// event types by wxWidgets are chosen dynamically and may not be the same
+// across runs or platforms.  We need an invariant name for each event type
+// of interest, which also makes the journal more legible than if we wrote mere
+// numbers.
+using Code = Identifier;
+
+struct Type;
+using Types = std::vector< Type >;
+
+// The list of event types to intercept and record.  Its construction must be
+// delayed until wxWidgets has initialized and chosen the integer values of
+// run-time event types.
+static const Types &TypeCatalog();
+
+// Lookup into the event type catalog during recording
+using ByTypeMap = std::map< wxEventType, const Type& >;
+static const ByTypeMap &ByType();
+
+// Lookup into the event type catalog during playback
+using ByCodeMap = std::unordered_map< Code, const Type & >;
+static const ByCodeMap &ByCode();
+
+// This sub-object populates the journal system's dictionary for playback
+struct RegisteredEventType : RegisteredCommand {
+   static bool DispatchEvent( wxArrayStringEx fields );
+   explicit RegisteredEventType( Code code )
+   : RegisteredCommand{ code.GET(), DispatchEvent }
+   {}
+};
+
+/*!
+An entry in a catalog describing the types of events that are intercepted
+and recorded, and simulated when playing back.
+*/
+struct Type : RegisteredEventType {
+
+   // Function that returns a list of parameters that, with the event type,
+   // are sufficient to record an event to the journal and recreate it on
+   // playback; or a nullopt for failure
+   using Serializer =
+      std::function< std::optional<wxArrayStringEx>( const wxEvent& ) >;
+
+   // Function that recreates an event at playback; or a null for failure
+   using Deserializer = std::function< std::unique_ptr<wxEvent>(
+      wxEventType, const wxArrayStringEx& ) >;
+
+   // Helper to keep casts out of the supplied function's definition
+   // Tag is the subtype of wxEvent expected by fn
+   template< typename Tag, typename SerialFn >
+   Serializer makeSerializer( SerialFn fn )
+   {
+      // Return an adaptor taking wxEvent
+      return [fn = std::move( fn )]( const wxEvent &e ){
+         return fn(static_cast<const Tag&>(e));
```

Comment:
I have very strong opinion that `static_cast` can't possibly be used within the template function. You leave both compiler and runtime helpless here.

It is really no difference between C-cast and `static_cast` in this case.

`dynamic_cast` is fast enough and this code is not performance-critical anyway.

Reply:
disagree

Reply:
static_cast is not as bad  C style pointer casts because it will not cast away const and will not allow conversion from base to derived when the inheritance relation is private or protected and is is not in the right scope for accessibility.

Reply:
If you tell me `static_cast` is absolutely never justified for pointer downcasting, I strongly disagree with that.  It is used in "type erasure" techniques such as you may soon review too for lib-xml.  I will refuse to "correct" the `static_cast`s in that PR unnecessarily.

wxWidgets event system is built around type erasure and is set up to invoke event handler functions without runtime checks.  See, for instance, this line of Audacity:

```
   EVT_BUTTON(ID_BUTTON_USECURRENTF, ContrastDialog::OnGetForeground)
```

Explore how the macro is implemented, if you dare, especially what `wxCommandEventHandler` means inside it.  Set a breakpoint in the function, arrange to hit it, and see the call stack.

This unchecked downcasting is pervasive in the handling of wxWidgets events and never a source of crashes in my experience.

wx/event.h even uses that same term, "type erasure:"

```
// The type-erased method signature used for event handling.
typedef void (wxEvtHandler::*wxEventFunction)(wxEvent&);
```

Reply:
Looks like an attempt to bypass compile time restrictions to me, with runtime overhead :)

Reply:
The simplistic opinion that "static_cast to downcast pointers is never right," is wrong.  I now know that is not Dmitry's opinion after the approving review of the lib-xml extraction.

It can be right when the code you need to examine to prove the correctness of the static_cast is not open ended.

But that isn't quite true with the wxWidgets events.  I think you can get undefined behavior from what should be properly written event handlers, but you have to work hard to do it.  Like construct your own wxEvent objects of the wrong class, and use SetEventType() on them, and then give them directly to a window to be processed.

It's possible but not usual.  It's deliberate "fraud" and not "accident."  SetEventType() isn't supposed to be used.  Events constructed by the libraries have the integer wxEventType consistent with the expected subtype of the wxEvent object.

Reply:
I like your style of debating :) I was not saying that it's right or wrong, I noted that you build an extra object at runtime to do something that looks like an attempt to persuade a compiler that you "know what you are doing"